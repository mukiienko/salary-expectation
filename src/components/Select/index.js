import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormControl from '@material-ui/core/FormControl';
import NativeSelect from '@material-ui/core/NativeSelect';

const Select = ({
  label,
  helper,
  value,
  valueName,
  disabled = false,
  handleChange,
  list = []
}) => (
  <FormControl>
    <InputLabel htmlFor={valueName}>{label}</InputLabel>
    <NativeSelect
      value={value}
      onChange={useCallback(e => handleChange(e.target.value))}
      disabled={disabled}
      input={<Input name={valueName} id={valueName} />}
    >
      {list.map(item => (
        <option key={item.id} value={item.id}>
          {item.name}
        </option>
      ))}
    </NativeSelect>
    <FormHelperText>{helper}</FormHelperText>
  </FormControl>
);

Select.propTypes = {
  label: PropTypes.string,
  helper: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  valueName: PropTypes.string,
  disabled: PropTypes.bool,
  handleChange: PropTypes.func,
  list: PropTypes.array
};

export default Select;
