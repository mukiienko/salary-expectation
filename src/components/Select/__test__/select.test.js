import React from 'react';
import renderer from 'react-test-renderer';
import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Select from '../index';

enzyme.configure({ adapter: new Adapter() });

const defaultProps = {
  handleChange: () => {},
  list: []
};
const SelectTest = props => <Select {...defaultProps} {...props} />;

describe('render Select', () => {
  it('render correctly select component', () => {
    const component = renderer.create(<SelectTest />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('props behaviour', () => {
    const props = {
      value: 0,
      label: 'label text',
      helper: 'helper text',
      disabled: true,
      list: [{ id: 0, name: 'TestNameOne' }, { id: 2, name: 'TestNameTwo' }]
    };
    const SelectTestWithProps = enzyme.mount(<SelectTest {...props} />);

    it('render correctly label props', () => {
      expect(
        SelectTestWithProps.find('label').contains(props.label)
      ).toBeTruthy();
    });

    it('render correctly helper props', () => {
      expect(SelectTestWithProps.find('p').contains(props.helper)).toBeTruthy();
    });

    it('render correctly disabled props', () => {
      expect(SelectTestWithProps.find('select').prop('disabled')).toBe(
        props.disabled
      );
    });

    it('render correctly list props', () => {
      expect(
        SelectTestWithProps.find('option')
          .at(0)
          .text()
      ).toBe(props.list[0].name);
      expect(
        SelectTestWithProps.find('option')
          .at(0)
          .props('value').value
      ).toBe(props.list[0].id);
      expect(SelectTestWithProps.find('option').length).toBe(props.list.length);
    });

    it('render correctly value props', () => {
      expect(SelectTestWithProps.find('select').prop('value')).toBe(
        props.value
      );
    });
  });

  describe('change behaviour', () => {
    const onChange = jest.fn(value => value);
    const handleChange = e => onChange(e);
    const props = {
      handleChange,
      valueName: 'testName',
      list: [{ id: 0, name: 'TestNameOne' }]
    };
    const SelectTestWithProps = enzyme.mount(<SelectTest {...props} />);

    it('fire change event', () => {
      const event = {
        preventDefault() {},
        target: { value: props.list[0].id }
      };
      SelectTestWithProps.find('select').simulate('change', event);
      expect(onChange).toHaveReturnedWith(props.list[0].id);
    });
  });
});
