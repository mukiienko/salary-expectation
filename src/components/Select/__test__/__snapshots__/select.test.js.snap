// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Select render correctly select component 1`] = `
<div
  className="MuiFormControl-root-1"
>
  <label
    className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
    data-shrink={false}
  />
  <div
    className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-formControl-44 MuiInput-formControl-31"
    onClick={[Function]}
  >
    <div
      className="MuiNativeSelect-root-23"
    >
      <select
        aria-invalid={false}
        className="MuiNativeSelect-select-24 MuiInputBase-input-53 MuiInput-input-38"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-60 MuiNativeSelect-icon-29"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <p
    className="MuiFormHelperText-root-69"
  />
</div>
`;
